source/fab/application.py

```diff
@@ -4,36 +4,50 @@
 # which you should have received as part of this distribution
 ##############################################################################
 from pathlib import Path
-from typing import Dict, Mapping, Type
+from typing import Dict, Type, Union, List
 
 from fab.database import SqliteStateDatabase
-from fab.language import Analyser
-from fab.language.fortran import FortranAnalyser, FortranWorkingState
-from fab.source_tree import ExtensionVisitor, FileInfoDatabase, TreeDescent
+from fab.language import Task, Command
+from fab.language.fortran import \
+    FortranAnalyser, \
+    FortranWorkingState, \
+    FortranPreProcessor
+from fab.source_tree import TreeDescent, ExtensionVisitor, FileInfoDatabase
 
 
 class Fab(object):
-    _extensions: Dict[str, Type[Analyser]] = {
-        '.F90': FortranAnalyser,
-        '.f90': FortranAnalyser
+    _extension_map: Dict[str, Union[Type[Task], Type[Command]]] = {
+        '.f90': FortranAnalyser,
+        '.F90': FortranPreProcessor,
     }
+    _command_flags_map: Dict[Type[Command], List[str]] = {}
 
-    def __init__(self, workspace: Path):
+    def __init__(self, workspace: Path, fpp_flags: str):
         self._state = SqliteStateDatabase(workspace)
-        self._extension_map: Mapping[str, Analyser] \
-            = {extension: analyser(self._state)
-               for extension, analyser in self._extensions.items()}
+        self._workspace = workspace
+        if fpp_flags != '':
+            self._command_flags_map[FortranPreProcessor] = (
+                fpp_flags.split()
+            )
 
     def run(self, source: Path):
-        visitor = ExtensionVisitor(self._extension_map)
+        visitor = ExtensionVisitor(self._extension_map,
+                                   self._command_flags_map,
+                                   self._state,
+                                   self._workspace)
         descender = TreeDescent(source)
         descender.descend(visitor)
 
         file_db = FileInfoDatabase(self._state)
         for file in file_db.get_all_filenames():
             info = file_db.get_file_info(file)
             print(info.filename)
-            print(f'    hash: {info.adler32}')
+            # Where files are generated in the working directory
+            # by third party tools, we cannot guarantee the hashes
+            if str(info.filename).startswith(str(self._workspace)):
+                print('    hash: --hidden-- (generated file)')
+            else:
+                print(f'    hash: {info.adler32}')
```

Comment:
This is not ideal but it is also throw-away code so not a problem.

Reply:
Agree it's not great, but after some time I couldn't find a way around it - the slightest difference in `GCC` version can cause differences in the generated output - in this case the version used by Github's Actions was adding 2 extra blank lines at the start of the produced file (and there's no way to control this it seems)... I think this is a nicer solution than trying to do something strange in the class (e.g. trimming out all leading blank lines until we              hit the first executable line of the original file, or something, which I considered doing as well)

Reply:
For the future we will probably want to examine the resuling files. The hashes are super useful to us internally but are not something we should be making users care about.